### PR TITLE
Add Gloo TCP_TLS transport

### PR DESCRIPTION
**Pull request by pbelevich**
This PR makes the following changes to root CMakeLists.txt:
1. If `USE_GLOO` is `ON` then try to build Gloo with OpenSSL headers(!) and set `USE_GLOO_WITH_OPENSSL` to `ON`
2. If `USE_GLOO_WITH_OPENSSL` is `ON` and the OS is Linux then try to find OpenSSL 1.1.x
3. If `OPENSSL_FOUND` then set `USE_TCP_OPENSSL_LOAD` to `ON` which means that Gloo will use OpenSSL headers to build TLS_TCP transport that is neither linked statically nor dynamically to OpenSSL but [tries to load it during runtime](https://github.com/facebookincubator/gloo/blob/2e0fe467aac7ed96a829548ec917edf6be8b1d84/gloo/transport/tcp/tls/openssl.cc#L4-L16)

@malfet @seemethere **I assume that during the release all CI release Linux environments will have OpenSSL 1.1.1(openssl-devel/libssl-dev) installed and it will automatically build PyTorch with TLS_TCP transport support.** 

**The remaining issue: mobile builds and libtorch builds have OpenSSL 1.0.2 which is not compatible with my code**

Depens on https://github.com/pytorch/builder/pull/708

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56442 Add Gloo TCP_TLS transport**

Differential Revision: [D27896285](https://our.internmc.facebook.com/intern/diff/D27896285)

### CODE REVIEW

File: .jenkins/pytorch/test.sh
Comment (pbelevich) [inline .jenkins/pytorch/test.sh:451]: put it at first place for testing purpose, will move it to the end after resolve all issues

File: CMakeLists.txt
Comment (pbelevich) [inline CMakeLists.txt:333]: Made it OPENSSL_init_ssl for testing purposes to make sure that we have libssl-dev/openssl-develop installed on CI environments

**Review by rohan-varma — COMMENTED** [review]
Are we planning to document `GLOO_DEVICE_TRANSPORT=TCP_TLS` (can be done in future PRs) can be set to enable tls transport?

File: .jenkins/pytorch/create_test_cert.sh
Comment (malfet) [inline .jenkins/pytorch/create_test_cert.sh:18]: Hmm, in order to make it easier for external users to run, can we add this logic to respective test in `test_c10d_gloo` using cryptography package (see https://cryptography.io/ )?
Something like:
with TemporaryDirectory() as temp_dir:
   from cryptography import x509
   subject_name = x509.Name([x509.NameAttributed(COUNTRY_NAME, "VA"),...])
   x509.CertificateBuilder().subject_name(subject_name).sign().....

File: torch/lib/c10d/GlooDeviceFactory.cpp
Comment (malfet) [inline torch/lib/c10d/GlooDeviceFactory.cpp:67]: I think function name is a bit misleading isn't it?
```suggestion
static std::string cstr_to_std_string(const char* c_str) {
```

File: torch/lib/c10d/GlooDeviceFactory.cpp
Comment (malfet) [inline torch/lib/c10d/GlooDeviceFactory.cpp:68]: nit
```suggestion
  return std::string (chars != nullptr ? chars : "");
```

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
LGTM, but please do not land until CI issues are resolved.

File: torch/lib/c10d/GlooDeviceFactory.cpp
Comment (pbelevich) [reply torch/lib/c10d/GlooDeviceFactory.cpp:68]: addressed

File: torch/lib/c10d/GlooDeviceFactory.cpp
Comment (pbelevich) [reply torch/lib/c10d/GlooDeviceFactory.cpp:67]: addressed

**Comment by seemethere** [comment]
I've went ahead and added `/opt/openssl` to the CI docker images so that this PR can pass the tests. Once that PR is merge you can rebase and re-run the test for these

File: CMakeLists.txt
Comment (malfet) [reply CMakeLists.txt:333]: Please remove required before merging the PR. Also, can you add a comment about why EXACT is needed here?

File: CMakeLists.txt
Comment (malfet) [inline CMakeLists.txt:283]: Can you please remove `AND LINUX` check from line 332 and move it here
```suggestion
cmake_dependent_option(
    USE_GLOO_WITH_OPENSSL "Use Gloo with OpenSSL. Only available if USE_GLOO is on." ON
    "USE_GLOO AND LINUX" OFF)
```

File: .jenkins/pytorch/create_test_cert.sh
Comment (malfet) [inline .jenkins/pytorch/create_test_cert.sh:15]: Regardless of this test, can we please have a test that setting `GLOO_DEVICE_TRANSPORT` to `FOOBAR` raises an exception (this way we can rely that this test will fail if PyTorch was compiled without TCP_TLS support)

File: .jenkins/pytorch/test.sh
Comment (malfet) [reply .jenkins/pytorch/test.sh:456]: Please make sure to limit it to just few configs.

File: .jenkins/pytorch/test.sh
Comment (malfet) [inline .jenkins/pytorch/test.sh:154]: Can we just run one or two subtests from the testsuite rather than a full one?

File: .jenkins/pytorch/test.sh
Comment (malfet) [reply .jenkins/pytorch/test.sh:468]: To elaborate on my previous comment, invoke this function if BUILD_ENVIRONEMNT equals to `pytorch_linux_xenial_py3_6_gcc7_test`

Comment (pbelevich) [reply .jenkins/pytorch/test.sh]: addressed

File: .jenkins/pytorch/create_test_cert.sh
Comment (pbelevich) [reply .jenkins/pytorch/create_test_cert.sh]: addressed

**Comment by pbelevich** [comment]
1. `find_package(OpenSSL REQUIRED)` removed from pytorch CMakeLists.txt
2. `USE_GLOO_WITH_OPENSSL` is OFF by default
3. Generating test cert/key in bash script was replaced with python code (create_test_cert.py), but I still have create_test_cert.sh to run it
4. I limited running TLS transport tests to `pytorch-linux-xenial-py3.6-gcc7-test` and `pytorch-linux-xenial-py3.6-gcc5.4-test`
5. I limited running TLS transport tests to only 44 tests in `ProcessGroupGlooTest` in `test_c10d_gloo.py`, it takes 20 seconds to run
6. I didn't add a test that checks that setting `GLOO_DEVICE_TRANSPORT` to `FOOBAR` raises an exception, because it requires some extra work, but I guarantee that `GLOO_DEVICE_TRANSPORT=FOOBAR` will crash with seg-fault(as expected)

**Comment by pbelevich** [comment]
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27896285).

**Comment by pbelevich** [comment]
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27896285).

**Comment by pbelevich** [comment]
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27896285).

**Comment by pbelevich** [comment]
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27896285).

**Pull request closed without merging**
